File: zombicideandroid/src/main/java/cc/game/zombicide/android/Stats.java
package cc.game.zombicide.android;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import cc.lib.utils.Reflector;
import cc.lib.zombicide.ZDiffuculty;
import cc.lib.zombicide.ZQuests;

public class Stats extends Reflector<Stats> {

    static {
        addAllFields(Stats.class);
    }

    private final Map<ZQuests, ZDiffuculty> completedQuests = new HashMap<>();

    public void completeQuest(ZQuests quest, ZDiffuculty difficulty) {
        completedQuests.put(quest, difficulty);
    }

    public boolean isQuestCompleted(ZQuests q, ZDiffuculty minDifficulty) {
        if (completedQuests.containsKey(q)) {
            return completedQuests.get(q).ordinal() >= minDifficulty.ordinal();
        }
        return false;
    }

    public Set<ZQuests> getCompletedQuests() {
        return new HashSet<>(completedQuests.keySet());
    }
}

File: zombicideandroid/src/main/java/cc/game/zombicide/android/Stats.kt
package cc.game.zombicide.android

import cc.lib.reflector.Reflector
import cc.lib.zombicide.ZDifficulty
import cc.lib.zombicide.ZQuests

class Stats : Reflector<Stats>() {
	companion object {
		init {
			addAllFields(Stats::class.java)
		}
	}

	private val completedQuests: MutableMap<ZQuests, ZDifficulty> = HashMap()

	fun completeQuest(quest: ZQuests, difficulty: ZDifficulty) {
		completedQuests[quest] = difficulty
	}

	fun isQuestCompleted(q: ZQuests, minDifficulty: ZDifficulty): Boolean {
		return completedQuests[q]?.let {
			it.ordinal >= minDifficulty.ordinal
		}?:false
	}

	fun getCompletedQuests(): Set<ZQuests> {
		return HashSet(completedQuests.keys)
	}
}